### PR TITLE
API auth gateway 

### PR DESCRIPTION
**Pull request by Tanwish2003**

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1,19 @@
+DATABASE_URL=postgres://new_auth_user:sqGE1invWBOJuuJVJDQDIsyVRxEqvM5F@dpg-cgrh93vdvk4n7bsbuea0-a.singapore-postgres.render.com/new_auth
```
Comment (akashg3627) [inline .env]: Put .env in .gitignore file

File: src/app.module.ts [hunk 2]
```diff
@@ -1,16 +1,38 @@
...
+import { JwtService } from '@nestjs/jwt';
+import { LoggerInterceptor } from './interceptor/LoggerInterceptor';
 @Module({
-  imports: [ConfigModule.forRoot(), DatabaseModule, UserModule],
+  imports: [
+    ConfigModule.forRoot(),
+    DatabaseModule,
+    AuthModule,
+    PassportModule,
+    // SequelizeModule.forRoot({
```
Comment (akashg3627) [inline src/app.module.ts]: Remove these

File: src/auth/auth.controller.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  async getProtectedResource(@Headers('authorization') authHeader: string): Promise<any> {
+    try {
+      const token = authHeader.split(' ')[1];
+      const { expired, payload } = await this.authService.verifyAccessToken(token);
+      if (expired) {
+        throw new UnauthorizedException(`Access token has expired`);
+      }
+      // Perform additional checks on the payload
+      // ...
+      return 'You have access to the protected resource';
```
Comment (akashg3627) [inline src/auth/auth.controller.ts]: return {message: "You have access ..."}

Comment (akashg3627) [reply src/auth/auth.controller.ts]: Always return json object as response, no string

File: src/auth/auth.service.ts [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+  constructor(
+    private config: ConfigService,
+    @Inject(USER_DAO)
+    private readonly userModel: typeof UserModel,
+    private jwtService: JwtService,
+  ) {
+    this.googleOAuth2Client = new OAuth2Client({
+      clientId: '603233410519-7l5m743sbl56ntteagmsortt1f32i2q7.apps.googleusercontent.com',
+      clientSecret: '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',
```
Comment (akashg3627) [inline src/auth/auth.service.ts]: put this in env and delete never commit it

File: src/auth/auth.service.ts [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+      redirectUri: `http://localhost:3000/v1/auth/login/google/callback`,
```
Comment (akashg3627) [inline src/auth/auth.service.ts]: make url as env, don't hardcode it

File: src/db/migrations/sql/init.sql [hunk 6]
```diff
@@ -1 +1,24 @@
-CREATE TABLE IF NOT EXISTS "user" ("id" VARCHAR(255) NOT NULL , "firstName" VARCHAR(255) NOT NULL, "lastName" VARCHAR(255) NOT NULL, "isActive" BOOLEAN NOT NULL DEFAULT false, "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL, "updatedAt" TIMESTAMP WITH TIME ZONE NOT NULL, PRIMARY KEY ("id"));
\ No newline at end of file
+CREATE TABLE IF NOT EXISTS "public"."user" (
+    "UserId" uuid NOT NULL,
```
Comment (akashg3627) [inline src/db/migrations/sql/init.sql]: id

File: src/db/migrations/sql/init.sql [hunk 7]
```diff
@@ -1 +1,24 @@
...
+    "createdAt" timestamptz,
+    "updatedAt" timestamptz,
+    "isVerified" bool DEFAULT false,
+    "hashedRT" varchar(255),
+    PRIMARY KEY ("UserId")
+);
+
+CREATE TABLE IF NOT EXISTS "public"."profile" (
+    "id" uuid NOT NULL,
+    "Name" varchar(255),
```
Comment (akashg3627) [inline src/db/migrations/sql/init.sql]: name

File: src/db/migrations/sql/init.sql [hunk 8]
```diff
@@ -1 +1,24 @@
...
+    "imageUrl" varchar(255),
+    "Metadata" json,
```
Comment (akashg3627) [inline src/db/migrations/sql/init.sql]: metadata

File: src/db/migrations/sql/init.sql [hunk 9]
```diff
@@ -1 +1,24 @@
-CREATE TABLE IF NOT EXISTS "user" ("id" VARCHAR(255) NOT NULL , "firstName" VARCHAR(255) NOT NULL, "lastName" VARCHAR(255) NOT NULL, "isActive" BOOLEAN NOT NULL DEFAULT false, "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL, "updatedAt" TIMESTAMP WITH TIME ZONE NOT NULL, PRIMARY KEY ("id"));
\ No newline at end of file
+CREATE TABLE IF NOT EXISTS "public"."user" (
```
Comment (akashg3627) [inline src/db/migrations/sql/init.sql]: Always use lower case in sql

File: src/db/models/Profile.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  })
+  Name: string;
+
+  @Column({ type: DataType.STRING }) imageUrl: Record<string, any>;
+
+  @Column({ type: DataType.JSON }) Metadata: Record<string, any>;
+
+  @ForeignKey(() => UserModel)
+  @Column(sequelize.UUID)
+  profileId: string;
```
Comment (akashg3627) [inline src/db/models/Profile.ts]: userId

File: src/db/models/Profile.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+  @BelongsTo(() => UserModel, 'UserId')
+  profile: UserModel;
```
Comment (akashg3627) [inline src/db/models/Profile.ts]: user

File: src/users/user.controller.ts [hunk 12]
```diff
@@ -1,30 +1,47 @@
-import { Body, Controller, Delete, Get, Inject, Param, Post } from '@nestjs/common';
```
Comment (akashg3627) [inline src/users/user.controller.ts:1]: delete if not needed

**Review by akashg3627 — COMMENTED** [review]
Please make these changes

File: src/db/migrations/sql/init.sql [hunk 13]
```diff
@@ -1 +1,39 @@
...
+CREATE TABLE "public"."profile"
+(
+  "id" uuid NOT NULL,
+  "Name" varchar(255),
+  "imageUrl" varchar(255),
+  "Metadata" json,
+  "profileId" uuid,
+  "createdAt" timestamptz NOT NULL,
+  "updatedAt" timestamptz NOT NULL,
+  "UserId" uuid,
```
Comment (akashg3627) [inline src/db/migrations/sql/init.sql]: userId

File: src/db/migrations/sql/init.sql [hunk 14]
```diff
@@ -1 +1,39 @@
...
+  CONSTRAINT "profile_UserId_fkey" FOREIGN KEY
+("UserId") REFERENCES "public"."user"
+("UserId") ON
+DELETE
```
Comment (akashg3627) [inline src/db/migrations/sql/init.sql]: ON DELETE CASCADE ON UPDATE CASCADE

File: src/db/migrations/sql/init.sql [hunk 15]
```diff
@@ -1 +1,39 @@
-CREATE TABLE IF NOT EXISTS "user" ("id" VARCHAR(255) NOT NULL , "firstName" VARCHAR(255) NOT NULL, "lastName" VARCHAR(255) NOT NULL, "isActive" BOOLEAN NOT NULL DEFAULT false, "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL, "updatedAt" TIMESTAMP WITH TIME ZONE NOT NULL, PRIMARY KEY ("id"));
\ No newline at end of file
+CREATE TABLE "public"."user"
```
Comment (akashg3627) [inline src/db/migrations/sql/init.sql]: CREATE TABLE IF NOT EXISTS

File: src/db/migrations/sql/init.sql [hunk 16]
```diff
@@ -1 +1,39 @@
...
+  "updatedAt" timestamptz,
+  "isVerified" bool DEFAULT false,
+  "hashedRT" varchar
+(255),
+  PRIMARY KEY
+("id")
+);
+
+
+CREATE TABLE "public"."profile"
```
Comment (akashg3627) [inline src/db/migrations/sql/init.sql]: CREATE TABLE IF NOT EXISTS

File: src/db/models/Profile.ts [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+    defaultValue: sequelize.UUIDV4,
+  })
+  id: typeof randomUUID;
+
+  @Column({
+    type: DataType.STRING(255),
+  })
+  name: string;
+
+  @Column({ type: DataType.STRING }) imageUrl: Record<string, any>;
```
Comment (akashg3627) [inline src/db/models/Profile.ts]: new line

File: src/db/models/Profile.ts [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+
+  @Column({ type: DataType.JSON }) metadata: Record<string, any>;
```
Comment (akashg3627) [inline src/db/models/Profile.ts]: new line

File: src/db/models/user.ts [hunk 19]
```diff
@@ -8,17 +9,34 @@ export class UserModel extends Model {
   @Column({
     primaryKey: true,
     allowNull: false,
-    type: DataType.STRING,
-    defaultValue: cuid,
+    type: sequelize.UUID,
+    defaultValue: sequelize.UUIDV4,
   })
-  id: string;
+  id: typeof randomUUID;
```
Comment (akashg3627) [inline src/db/models/user.ts:15]: string is ok

File: .env (hunk 1 above)
Comment (akashg3627) [reply .env]: don't commit .env file, add it to gitignore

File: src/db/models/Profile.ts [hunk 20]
```diff
@@ -0,0 +1,33 @@
+import { Column, DataType, Model, Table, ForeignKey, BelongsTo } from 'sequelize-typescript';
```
Comment (akashg3627) [inline src/db/models/Profile.ts:1]: file name profile.ts

File: src/auth/auth.service.ts [hunk 21]
```diff
@@ -0,0 +1,276 @@
...
+      sub: id,
+      email: email,
+    };
+
+    const [at, rt] = await Promise.all([
+      this.jwtService.signAsync(jwtPayload, {
+        secret: this.config.get<string>('AT_SECRET'),
+        expiresIn: '7d',
+      }),
+      this.jwtService.signAsync(jwtPayload, {
```
Comment (akashg3627) [inline src/auth/auth.service.ts:189]: refresh token is not a jwt, its just a random string

File: src/db/migrations/sql/init.sql (hunk 15 above)
Comment (Tanwish2003) [reply src/db/migrations/sql/init.sql]: we tried doing this but invalid synatx is bieng displayed there while we make the change

File: src/auth/auth.service.ts (hunk 21 above)
Comment (Tanwish2003) [reply src/auth/auth.service.ts:189]: will it be okay if we keep it a jwt , cause we wrote the whole logic considering that , and it is part of many routes

File: src/db/models/user.ts (hunk 19 above)
Comment (Tanwish2003) [reply src/db/models/user.ts:15]: it was not accepting string , in the code you provided cuid was used , but its depricated now , so we switched to uuid and it was throwing error while using string

**Review by akashg3627 — APPROVED** [review]

**Pull request merged**